src/storage/patch/N3Patcher.ts

```diff
@@ -0,0 +1,150 @@
+import type { Readable } from 'stream';
+import { newEngine } from '@comunica/actor-init-sparql';
+import type { ActorInitSparql } from '@comunica/actor-init-sparql';
+import type { IQueryResultBindings } from '@comunica/actor-init-sparql/lib/ActorInitSparql-browser';
+import { Store } from 'n3';
+import type { Quad, Term } from 'rdf-js';
+import { mapTerms } from 'rdf-terms';
+import { Generator, Wildcard } from 'sparqljs';
+import type { SparqlGenerator } from 'sparqljs';
+import { BasicRepresentation } from '../../http/representation/BasicRepresentation';
+import { isN3Patch } from '../../http/representation/N3Patch';
+import type { N3Patch } from '../../http/representation/N3Patch';
+import type { Representation } from '../../http/representation/Representation';
+import { RepresentationMetadata } from '../../http/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../../http/representation/ResourceIdentifier';
+import { getLoggerFor } from '../../logging/LogUtil';
+import { INTERNAL_QUADS } from '../../util/ContentTypes';
+import { ConflictHttpError } from '../../util/errors/ConflictHttpError';
+import { InternalServerError } from '../../util/errors/InternalServerError';
+import { NotImplementedHttpError } from '../../util/errors/NotImplementedHttpError';
+import { readableToQuads } from '../../util/StreamUtil';
+import type { RepresentationPatcherInput } from './RepresentationPatcher';
+import { RepresentationPatcher } from './RepresentationPatcher';
+
+/**
+ * Applies an N3 Patch to a representation, or creates a new one if required.
+ * Follows all the steps from Solid, §5.3.1: https://solid.github.io/specification/protocol#n3-patch
+ */
+export class N3Patcher extends RepresentationPatcher {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly engine: ActorInitSparql;
+  private readonly generator: SparqlGenerator;
+
+  public constructor() {
+    super();
+    this.engine = newEngine();
+    this.generator = new Generator();
+  }
+
+  public async canHandle({ patch }: RepresentationPatcherInput): Promise<void> {
+    if (!isN3Patch(patch)) {
+      throw new NotImplementedHttpError('Only SPARQL update patches are supported');
+    }
+  }
+
+  public async handle(input: RepresentationPatcherInput): Promise<Representation> {
+    const patch = input.patch as N3Patch;
+
+    // No work to be done if the patch is empty
+    if (patch.deletes.length === 0 && patch.inserts.length === 0 && patch.conditions.length === 0) {
+      this.logger.debug('Empty patch, returning input.');
+      return input.representation ?? new BasicRepresentation([], input.identifier, INTERNAL_QUADS, false);
+    }
+
+    if (input.representation && input.representation.metadata.contentType !== INTERNAL_QUADS) {
```

Comment:
`canHandle`?

Reply:
The usual semantic discussion of what this class is specifically responsible for. 😉 But either is fine here for me so I'll move it.